Improve the hackage IA.

cskksc opened the pull request:
**What does your Instant Answer do?**
Fetches packages or methods based on signatures for Haskell.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Instant access to package names and versions. Or method signatures.

**What is the data source for your Instant Answer? (Provide a link if possible)**
https://www.haskell.org/hoogle/
The hackage api doesn't support json responses for search operations.

**Why did you choose this data source?**
It provides accurate information about packages and their versions. Is widely used in the haskell community.

**Are there any other alternative (better) data sources?**
[Hayoo](http://hayoo.fh-wedel.de/) is also comparable. But the library version information is missing. So I think [Hoogle](https://www.haskell.org/hoogle/) is better.

**What are some example queries that trigger this Instant Answer?**
```hask json```, ```hask [a]-> a```, ```map hask```

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Haskell programmers.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
Not sure. There's one specific for Hayoo and not generic haskell packages. [Idea](https://duck.co/ia/view/hayoo)

**Which existing Instant Answers will this one supersede/overlap with?**
hackage, hayoo. #1950 

**Are you having any problems? Do you need our help with anything?**
Not right now.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
No authorization required. And its free.

**Where did you hear about DuckDuckHack? (For first time contributors)**
Friends

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
- Package listing:
  ![package](http://f.cl.ly/items/0j02001S022T0w3p0U2K/package.png)

- Method signature
  ![method_sig](http://f.cl.ly/items/110n0o3g1S1R2s1P3Q1e/method_sig.png)
## Checklist
Please place an 'X' where appropriate.

```
[] Added metadata and attribution information
[x] Wrote test file and added to t/ directory
[x] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[x] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[x] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [x] Google Chrome
        [x] Firefox
        [] Opera
        [x] Safari

    - iOS (iPhone)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [x] Firefox
        [] Native Browser
        [x] Google Chrome
        [] Opera
```

moollaza commented:
>Which existing Instant Answers will this one supersede/overlap with?
hackage, hayoo. #1950 

@cskksc I'm not a Haskell user, so I'm not quite sure which of these IA's/Sources are necessary. Do you think with this IA merged, the others are still useful? Should we remove them?

cskksc commented:
The functionality offered by this is very similar to the Hayoo IA. But IMHO the results from Hoogle are better than Hayoo. For example, [hayoo list](https://duckduckgo.com/?q=hayoo+list) vs [hask list](http://twist-joker.codio.io:5000/?q=hask+list&ia=hackage). Also, Hoogle returns version numbers for packages which would allow us to sort them. Plus hoogle is a part of haskell.org. Having an api for the actual [Hackage](https://hackage.haskell.org/) site would have been nice. But its search doesn't support json responses, only HTML. https://hackage.haskell.org/api#search .

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,97 @@
+                    url: item.location
+                };
+                if (v) {
+                    info.subtitle = "Version: " + v;
```
please switch to `info.altsubtitle` this will prevent the space between the title and subtitle

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,97 @@
+            templates: {
+                group: 'text',
+                detail: false,
+                item_detail: false
```
Let's define some variants to format things a little better:

```js
templates: 
  variants: {
    tileTitle: "2line",
    tileSnippet: "large"
  }
}
```

moollaza commented:
@cskksc this is off to a great start. I see now that we're parsing the "docs" portion of the result to get what we need -- is there no API endpoint with better structured data? Can we ask the API maintainers to improve the API for us?

Also, I see a few results have no `docs` string -- perhaps we should skip these results so we don't a mostly blank tile. Alternatively we could put `-` for the Version and `No description available` so it's more clear why it's missing.

@abeyang any thoughts on design? We don't have anything more available to use beside the title, version and snippet it seems.

cskksc commented:
@moollaza @abeyang Made the changes to the ui. And skipped the results with empty ```docs``` field. This is how it looks now,

<img width="1000" alt="ui" src="https://cloud.githubusercontent.com/assets/5201497/8506766/f73a659a-2242-11e5-8230-60261766cb7c.png">

I have also raised an [issue](https://github.com/ndmitchell/hoogle/issues/120) on the hoogle repository. Lets see if we can get any help in improving the api.

abeyang commented:
@cskksc Is there any more info provided by the API, such as number of downloads or last updated? Would be nice to have it more consistent with other similar IAs:

![sublime-v1 1](https://cloud.githubusercontent.com/assets/794084/8550718/27ede48c-2487-11e5-8cd8-221c07cc79ea.jpg)

cskksc commented:
@abeyang Not right now. But the API creator has been very helpful and confirmed that they are working on a release of Hoogle 5.0 which would have an extensive API. Then we can further improve the results for packages and method signatures both. I'm not sure about the release date. https://github.com/ndmitchell/hoogle/issues/120

abeyang commented:
@cskksc Ok, if that's the case, then instead of using `.tile__content--lg` (associated with `.tile__content`), can we use `.tile__content--m`? There's too much empty space under the snippets, as it currently is.

cskksc commented:
@abeyang Yeah that looks better. I couldn't find any documentation on how to assign css classes to html elements. Could you point me to some resources ?

moollaza commented:
>I couldn't find any documentation on how to assign css classes to html elements. Could you point me to some resources ?

@cskksc sorry about that! It's a little bit buried in the docs but we cover it here: https://duck.co/duckduckhack/templates_reference#directly-specifying-classes

If you look through the repo for `elClass` you'll be able to find other working examples.

---

However, the usual way to set the class for elements like the snippet are to use the Variants:

```js
variants: {
   tileSnippet: 'medium'
}
```

But it seems we haven't implemented the "medium" option for some reason. So setting that with `elClass` will work.

We also have an optional `snippetChars` property you can use to control the number of characters allowed in the snippet, but those classes will determine the height of the snippet div.

cskksc commented:
@moollaza @abeyang Ah, my bad. I thought elClass is only used for text & colours. Now I've applied the ```tile__content--m``` class. Looks like this.

<img width="795" alt="screen shot 2015-07-14 at 4 51 35 pm" src="https://cloud.githubusercontent.com/assets/5201497/8672353/110aa010-2a4b-11e5-8974-91dc3f4b90c4.png">

moollaza commented:
@cskksc terribly sorry, but it looks like @abeyang was mistaken about the existence of a `tile__content--m` class. We don't actually have one, but in modifying the `--lg` class on the element, it made it look like it existed as the height changed.

It looks like we'll need to set an explicit height for the tiles via CSS. Sorry for the back and forth but can you set the tiles to a height of `10em` using CSS?

jagtalon commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,108 @@
+
+
+        // limit display results to 30
+        var results = api_result.results;
```
Check if `api_result` exists first before accessing `api_result.results`.

```javascript
if(!api_result || api_result.error) {
    return Spice.failed('hackage');
}
```

jagtalon replied:
Also, we should probably check if `results` has a length greater than 0

jagtalon commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,108 @@
+        var source = $(script).attr("src");
+        var query = source.match(/hackage\/([^\/]*)/)[1];
+
+        if (api_result.error) {
```
Does the API actually return an `error` property if things don't go well? Also, see https://github.com/duckduckgo/zeroclickinfo-spice/pull/1977/files#r34880825.

jagtalon commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,108 @@
+            }
+            while(v.indexOf(".") != -1) {
+                v = v.replace(".","");
+            }
```
What about `"1.2.3".replace(/\./g, "")`?

jagtalon commented on lib/DDG/Spice/Hackage.pm:
```diff
@@ -0,0 +1,16 @@
+
+triggers startend => 'hask';
+
+spice to => 'http://www.haskell.org/hoogle/?mode=json&hoogle=$1';
```
This redirects to `https://www.haskell.org/hoogle/?mode=json&hoogle=` (note the HTTPS part). Our Nginx config doesn't handle redirects at the moment, so we can't have any endpoints that redirect. Please change "HTTP" to "HTTPS".

jagtalon commented on lib/DDG/Spice/Hackage.pm:
```diff
@@ -0,0 +1,16 @@
+
+use DDG::Spice;
+
+triggers startend => 'hask';
```
Trigger on "hackage", "haskell", "hayoo" or "hoogle"?

jagtalon commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,108 @@
+            data: results,
+            meta: {
+                sourceName: "Hackage",
+                sourceUrl: "https://hackage.haskell.org/packages/search?terms=" + query,
```
Shouldn't this link to `https://www.haskell.org/hoogle/?hoogle=` since we're using Hoogle?

jagtalon commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,108 @@
+                v = v.replace(".","");
+            }
+            // Assuming semantic version groups do not exceed 4
+            toAppend = 4 - v.length;
```
When declaring a new variable, add `var`.

cskksc commented:
@moollaza @abeyang Thats perfectly fine. It did seem that ```tile__content--m``` change had that effect on the sizes :smile: . I've changed the height to ```10em``` now. Also added a chars limit of 85. Looks like this,

<img width="912" alt="screen shot 2015-07-17 at 4 38 31 pm" src="https://cloud.githubusercontent.com/assets/5201497/8746002/be65ee8c-2ca2-11e5-8b79-a018aa5eaa2e.png">

cskksc commented:
@jagtalon I've made the changes according to your comments. The ```hask``` trigger was temporary to avoid conflicts with [Hayoo IA](https://github.com/duckduckgo/zeroclickinfo-spice/blob/master/lib/DDG/Spice/Hayoo.pm#L15).

cskksc commented:
@abeyang ping.

moollaza commented on share/spice/hackage/hackage.css:
```diff
@@ -0,0 +1,3 @@
+.tile__body {
+    height: 10em;
```
@cskksc looks like this is causing problems? It's cutting off the bottom of the tile for me:

![hackage_json_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9174864/975b2846-3f50-11e5-9104-e7a33f1debfc.png)

Removing the CSS causes this to look OK:

![hackage_json_at_duckduckgo_and_1____projects_zeroclickinfo-goodies__ssh_](https://cloud.githubusercontent.com/assets/873785/9174884/b13ee9fa-3f50-11e5-811e-d7f2685151e6.png)

cskksc commented:
@moollaza Yeah, it looks bad. Was looking ok on my machine. Have removed the CSS.

moollaza commented:
@cskksc thanks!

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+            meta: {
+                sourceName: "Hoogle",
+                sourceUrl: "https://www.haskell.org/hoogle/?hoogle=" + query,
+                snippetChars: 85
```
Lets increase this to `180`, there's lot more text we can show!

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+                detail: false,
+                item_detail: false,
+                variants: {
+                    tileTitle: "2line"
```
let's set a large variant for the snippet so the height of the tiles is consistent:

`tileSnippet: "large"`

cskksc replied:
It was ```large``` before, but @abeyang suggested that it should be medium because of the empty space in many tiles. So we wanted  ```medium``` tile-height, but there's no CSS class for it. Then we settled on using ```height: 10 em```. But having both the ```height``` and ```snippetChars``` seems to be causing the text to cut off like that, so I have removed the css. So only ```snippetChars: 170``` should be used then ?

moollaza replied:
@cskksc sorry for the confusion. Let's go with the "Large" Snippet, and "170" char snippet. That way the text fills up more space and we don't have consistent gaps below the snippets.

/cc @abeyang

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+            normalize : function(item) {
+                var v = version(item.docs);
+                var description = docString(item.docs);
+                if (description == "") {
```
Seems like this might not be working correctly? Let's make sure the description also contains more than 1 character:

Looks like it has no description:
![hackage_math_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9177894/23ca165a-3f62-11e5-8a9b-9986c4e600ad.png)

Description is too short:
![hackage_json_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9177906/32cff5b6-3f62-11e5-960a-ef8c3ead974d.png)

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+            return Spice.failed('hackage');
+        }
+        var results = api_result.results;
+        if (results.length > 30)
```
Any chance there's an API param that will limit us to 30 results?

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+    env.ddg_spice_hackage = function(api_result) {
+
+        // get documentation string
+        function docString(doc) {
```
Let's define these private functions outside the callback, above `env.ddg_spice_hackge`. We're working in a closure so they'll be accessible in the `ddg_spice_hackge` but won't get added to the global `env` :smiley:

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+        // get version from docs.
+        // supported version formats: [1.2, 1.2.3, 1.2.3.4]
+        function version(doc){
+            try {
```
I think we can do away with the `try..catch` block here:

```js
var version = doc.match(/\d\.\d\.?\d?\.?\d?/);
return version || undefined;
```

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+        // converts "0.0.1" to 10
+        // helps in comparing version strings in different format.
+        function versionInt(v) {
+            if (v == undefined) {
```
Let's use `===`

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,102 @@
+            if (v == undefined) {
+                return -1;
+            }
+            v = v.replace(/\./g,"")
```
Doesn't seem like we need a regex here to match `"."`?

moollaza replied:
Ah nevermind, I see the `/g`. :+1:

cskksc commented:
@moollaza hey, I've made the other changes, have removed ```snippetChars``` and set tile height to small for now.

moollaza commented:
@cskksc we actually wanted the taller tile (`"large"`) and higher charsnippet (`175`) so we can show more! Let's go with that :smiley:

cskksc commented:
@moollaza Okay, now tile snippet size is ```large```, & char limit is ```170```. It did fill up the empty space. Looks like this,
<img width="1146" alt="screen shot 2015-08-11 at 9 05 12 am" src="https://cloud.githubusercontent.com/assets/5201497/9189656/511df94e-4009-11e5-921e-54d2df2e80e7.png">

abeyang commented:
Looks good :+1: . Can we change the tab title to something more generic, like "Package" or "Software"?

cskksc commented:
@abeyang Cool :smiley:. Changed it to "Software", as not all results would be packages.

moollaza commented on lib/DDG/Spice/Hackage.pm:
```diff
@@ -0,0 +1,16 @@
+
+use DDG::Spice;
+
+triggers startend => 'haskell', 'hoogle', 'hackage';
```
Wondering if we should add "haskell packge"? I'm not sure if users are likely to have the word "package" in their query or not.

Alternatively we could strip it out of the remainder if think it's going to be there.

cskksc replied:
Lets remove words "package" or "packages" from the query to be safe.

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,94 @@
+// converts "1.2.3" to 1230
+// converts "0.0.1" to 10
+// helps in comparing version strings in different format.
+function versionInt(v) {
```
Whoops, these private functions need to be inside our closure (`(function(env) { ... } ()`), but outside our callback (`env.ddg_spice_hackage`).

Let's move them down, so they aren't declared in the global namespace.

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,94 @@
+    "use strict";
+    env.ddg_spice_hackage = function(api_result) {
+
+        if(!api_result || api_result.results.length == 0) {
```
Should probably add `!api_result.results ||` in the middle here.

moollaza commented on share/spice/hackage/hackage.js:
```diff
@@ -0,0 +1,94 @@
+                var v = version(item.docs);
+                var description = docString(item.docs);
+                if (description === "" || description.length < 2) {
+                    return;
```
this should be `return null`

moollaza commented:
@cskksc this is looking great. I've made some final comments and then this should be good for merge! Thanks for doing a tremendous job :+1:

cskksc commented:
Thanks for all the help @moollaza. My first open source contribution has been to DDG and the experience has been great ! :smiley:

moollaza commented:
@cskksc sorry for the delay! This LGTM :+1:

cskksc commented:
Cool :smile:. Right now, Hayoo IA and this one shares a trigger, "haskell". This won't cause any problems right ?

moollaza commented:
@cskksc do we still need hayoo? Is it better when it comes to returning results for the "haskell" trigger? We could narrow the triggers for hayoo, or let it only trigger on "hayoo"?

cskksc commented:
@moollaza IMO "hackage" source is better, but I'm not a regular haskell user so I can't say for sure. May be we can take some feedback on the haskell mailing list, after this IA is live. I'm changing the trigger to only "hayoo" for now. If both IA's are working simultaneously, on what parameters would one get preference ? latency, no of results etc ?

cskksc commented:
Follow up [PR](https://github.com/duckduckgo/zeroclickinfo-spice/pull/2121).

moollaza commented:
@cskksc to answer your question:

>If both IA's are working simultaneously, on what parameters would one get preference ? latency, no of results etc ?

It's a race for Spices. Whichever calls `Spice.add()` first get's to display first, assuming nothing else, more important also tries to show an answer
